### DISCUSSION TITLE
Good Stuff.

### DISCUSSION THREAD

**Discussion by panoply** [FIameCaster/prism-code-editor#21] [General]
@FIameCaster 

This is a well done lib. I was looking to do something similar with PrismJS but never got around to exploring things at depth. I'm interested on the per-line tokenizing tactic you've got in place. Initial glance, there is good foundation and extending the existing logic while keeping things snappy beyond 1k LoC is something I'd be interesting in exploring as it's both a challenging and an itch I've been wanting to scratch with Prism.

Given you've already delved into the lexing algorithm of PrismJS, I'd be curious to hear your thoughts around the incremental updates, your experience in refactors to get it where it is now and any issues you encountered. One of the biggest perf-hits with PrismJS is the grammars and it's tokenization method. In my previous attempts, getting Prism to behave reactively in 1k - 2k LoC is difficult. 

The per-line targeting of syntactics really peaked my interest.

**Comment by jonpyt** [comment]
Hi @panoply. Thanks for showing interest!

The tokenization algorithm here is very similar to the one in PrismJS with only minor changes. Splitting into lines is done after the tokenization. To avoid losing context when splitting, tags must be closed and immediately opened again when a line break is found. You can see the logic for it [here](https://github.com/FIameCaster/prism-code-editor/blob/3.3.2/package/src/prism/core.js#L74). Splitting the HTML into lines allows diffing when updating the DOM instead of replacing all the HTML like what's done in many other code editors using PrismJS or Highlight.js for syntax highlighting. Here's the [diffing algorithm](https://github.com/FIameCaster/prism-code-editor/blob/3.3.2/package/src/core.ts#L74).

The only behavioral change I've made to the tokenization algorithm is skipping empty matches. Empty matches can cause an infinite loop which is why PrismJS has [this failsafe](https://github.com/PrismJS/prism/blob/v1.29.0/components/prism-core.js#L957). By skipping empty matches, an infinite loop becomes impossible, which means we can remove this failsafe and also the length of the linked list since it's unused.

I also rewrote the tokenization algorithm slightly to use a singly linked list instead of a doubly linked list, but this optimization doesn't change the behavior of the tokenizer at all.

Keeping things snappy beyond 1k LoC would require both virtualization and incremental parsing. Virtualization is something I've avoided doing since it would massively increase the complexity of nearly everything, defeating the purpose of this library IMO.

Regarding incremental tokenization, I'd say it's nearly impossible. Prism's regex based grammars are very simple and flexible. Doing incremental parsing with them would be anything but simple, and not worth it.

**Comment by panoply** [comment]
Thanks for getting back to me on this. 

I really like what you've done here. I started something similar some time ago which is still very much WIP (see [papyrus.js.org](https://papyrus.js.org/#!/liquid)). I found it interesting to see the common approaches (repo: https://github.com/panoply/papyrus) taken, but of course papyrus is not nearly as fluid and well composed as PCE.I think there is great need for a solution of this nature. 

I wanted to see about implementing the work you've done in PCE for Papyrus but currently it is not viable, mainly because of the underlying approach I've taken so as the solution can work within node environments, along with rendering logic and extended grammars I have in place (which are solely for my personal tastes). 

> Keeping things snappy beyond 1k LoC

FWIW: I was able to have 2k LoC and things remained snappy in pce. 

>  Prism's regex based grammars are very simple and flexible. Doing incremental parsing with them would be anything but simple, and not worth it.

Indeed.

**Reply by jonpyt** [reply]
Nice project!

I can definitely tell that it's a WIP. I'd like to point out some apparent issues. I'm sure you know most of them.

- Horizontal scrolling is rather janky due to slow layout reflows that seem to be caused by [this statement](https://github.com/panoply/papyrus/blob/master/src/modes/editor.ts#L197) setting the width of the active line highlight.
- The active line highlighting is often incorrect when selecting multiple lines of text.
- The active line highlight can disappear when typing. I'm guessing `morphdom` is removing the `active` class.
- You're using `document.execCommand` to insert text. One issue with this is that Chrome (and to a lesser degree Safari) dispatches two input events for every line of inserted text. Selecting a 600 line document and indenting with the tab key led to a 40 second freeze as the document was highlighted over 1200 times.
- You're using `morphdom` to do diffing. It's rather effective for reducing style and layout calculations since it identifies precisely which text nodes have changed, but it does have quite high overhead. It has to parse the HTML to create temporary DOM nodes that are diffed against the current document. From my testing it performs slightly better than replacing the entire document, but not by much. It's still multiple times slower than my text based diffing for normal typing.

> I wanted to see about implementing the work you've done in PCE for Papyrus but currently it is not viable, mainly because of the underlying approach I've taken so as the solution can work within node environments, along with rendering logic and extended grammars I have in place (which are solely for my personal tastes).

The Prism instance I'm using runs just fine on Node.js. I can see that you're supporting rendering on Node.js to an HTML string. I don't see why my Prism instance and diffing couldn't work for you, but you would need to re-implement `Prism.highlight`.

> FWIW: I was able to have 2k LoC and things remained snappy in pce.

2k lines is right on the limit for me too. But with 10s of thousands of nodes in the DOM, the whole website slows down which isn't great. With the same document in Papyrus I'd describe the experience as unpleasant.

You have a lot of work ahead of you, but once you polish the rough edges, it will definitely be better than most of the lightweight code editors out there. I hope I've given you some tips on where to start.

**Reply by panoply** [reply]
Hey, 

Yes, very much WIP and all issues I am aware of (it's merely a hack). PCE would be a much better fit overall. The main use-case for Papyrus is for Æsthetic documentation (see: https://github.com/panoply/esthetic/tree/latest). The documentation logic can be found here (https://aesthetic.js.org/rules/markup/forceAttribute/). There are multiple instances of it on a per-page basis, given that it is used to describe different rules the formatter supports etc. 

This node implementation is key here. One of the aspects with PCE is that `document` is exposed in global scope, the overall architecture makes things a tad verbose.

Is there some consideration to be had with existing API? Would you be open for PR/Suggestions?

**Reply by jonpyt** [reply]
> This node implementation is key here. One of the aspects with PCE is that document is exposed in global scope, the overall architecture makes things a tad verbose.

What would be possible is creating a seperate entry point, for example `prism-code-editor/ssr` that exports a function to render an editor as an HTML string. Only the `insertSpaces` option would not automatically be encoded in the string. It could default to `false` if any tab characters are found in the document, but it's probably better to encode it in a data attribute.

Then we'd have another entry point `prism-code-editor/hydrate` exporting a function that making the editor interactive by attaching event listeners. It also needs to iterate over each line to populate the `prevLines` array for [the diffing](https://github.com/FIameCaster/prism-code-editor/blob/3.3.3/package/src/core.ts#L83) to work. Lastly, it must reconstruct the options used to create the editor by reading the DOM. Otherwise the code for the hydrate function would largely be a copy-paste from `createEditor`.

There could also be a `hydrateRoot` function that takes in a `root` and calls `root.getElementsByClassName("prism-code-editor")`, hydrates all of the elements, and returns an array of `PrismEditor` objects.

To handle extensions, the hydrate functions could receive a callback function. This callback function would receive the constructed options and return an array of extensions.

Hydrating an element multiple times would be catastrophic, so we must handle this. We could add a data attribute to hydrated editors and throw an error if that attribute is present. A better solution is to make a `WeakMap` that maps the element to it's corresponding `PrismEditor` object. If the element already has an entry in that map, return its associated editor object instead. The `createEditor` function must also write to this map.

Lastly, how could we handle the `onTokenize`, `onUpdate`, and `onSelectionChange` options? It's obviously impossible to encode a function in an HTML string. Should these just not be supported at all?

### Necessary changes

> One of the aspects with PCE is that document is exposed in global scope

There are obviously some changes to the current library that must be made. The [template element](https://github.com/FIameCaster/prism-code-editor/blob/3.3.3/package/src/core.ts#L293) could be assigned in the `createTemplate` function with `templateEl ||= document.createElement("div")`.

The global [`selectionchange` handler](https://github.com/FIameCaster/prism-code-editor/blob/3.3.3/package/src/core.ts#L340) could be replaced with a function like this:

```ts
let added: boolean
let selectionChange: undefined | (() => void)

export const setSelectionHandler = (handler?: () => void) => {
  if (!added) {
    document.addEventListener("selectionchange", selectionChange?.())
    added = true
  }
  selectionChange = handler
}
```

### Final notes

Something I'd like to know is if this SRR/hydration API would be easy to integrate with static site generators like Eleventy or Astro and fullstack frameworks like Next.js, Nuxt, SvelteKit, etc.. If not, then how?

I like this idea, but it's maybe best suited for `v4.0.0`. I want to work on #18 first though.

I'd love to hear your thoughts on this.

**Reply by panoply** [reply]
Hydration can be achieved a few different ways here. It's relatively inexpensive to perform in a build environment, different question on the server.  I'll respond in better detail tomorrow, but one _possible_ route might be to leverage the DOM. Given that the tree will require traversal during hydration, it might an idea to provide a structure via annotation.

Looking forward to thinking more deeply about this.

**Reply by jonpyt** [reply]
> It's relatively inexpensive to perform in a build environment, different question on the server.

Constructing the HTML string for an editor will be very fast. Shouldn't impact server response times during SSR much at all.

> it might an idea to provide a structure via annotation.

I don't think I understand why we need any annotations in the DOM.

Anyway, I looked into using this API with Astro, and it looks very promising.

```astro
---
import { renderEditor } from "prism-code-editor/ssr"
import "prism-code-editor/prism/languages/tsx"
import "prism-code-editor/layout.css"
import "prism-code-editor/scrollbar.css"
import "prism-code-editor/themes/github-dark.css"

const editor = renderEditor(Astro.props)
---

<Fragment set:html={editor} />

<script>
  import { mountAll } from "prism-code-editor/client"
  import "prism-code-editor/prism/languages/tsx"
  
  const editors = mountAll(document.body, options => [
    // Extensions here
  ])
</script>
```

This uses the [`set:html` directive](https://docs.astro.build/en/reference/directives-reference/#sethtml) to display the editor and a [client-side script](https://docs.astro.build/en/guides/client-side-scripts/#client-side-scripts) to do the hydration. With a [`Fragment`](https://docs.astro.build/en/basics/astro-syntax/#fragments) we don't even need a wrapper element to add the HTML to.

When using this Astro component, there would be zero layout shifts. You can even do the hydration in a dynamic import, so almost zero JavaScript gets downloaded before the page loads.

**Reply by panoply** [reply]
> encode it in a data attribute
> reconstruct the options used to create the editor by reading the DOM. 
> We could add a data attribute to hydrated editors 

Referring to these operations when I mentioned annotation. Hydration need additional contexts, passing data within attributes makes the most sense.  

> To handle extensions, the hydrate functions could receive a callback function.
> Lastly, how could we handle the onTokenize, onUpdate, and onSelectionChange options? It's obviously impossible to encode a function in an HTML string

This is where things will get a little tedious imo, but maybe some restrictions are good for the sake of sanity. Though there could be situations where developers set multiple instances with each instance differing in the features and capabilities to apply. If that route was to be taken, the only slightly possible way would be to compose a object representation and possibly leverage the LZ-based compression algorithm, but it feels really "hacky"...

**Reply by panoply** [reply]
RE https://github.com/FIameCaster/prism-code-editor/discussions/21#discussioncomment-10078012

I've not used Astro and when it comes to the virtual DOM I am leveraging a different full stack web framework which is a successor to mithril.js and is not yet public (the author is still working on some things) so my base knowledge of execution and how these frameworks are handling Hydration is limited. Hydration in the project I am referring to is elegantly handled by walking the tree, comments are used to signal for component hydration, which might be worth considering, though I don't know how viable it would be for PCE.

I am however well versed with 11ty and things are rather simple in that context.

**Reply by jonpyt** [reply]
> Referring to these operations when I mentioned annotation. Hydration need additional contexts, passing data within attributes makes the most sense

The `textContent` of the editor becomes the `value` option. `tabSize` is already added as an inline style. The `language`, `lineNumbers`, `readOnly`, and `wordWrap` options can all be determined by reading the class-name of the scroll container. Only `insertSpaces` needs an annotation as I mentioned earlier. A data attribute for it would work just fine.
